pkg/endpoint/endpoint.go

```diff
@@ -1506,19 +1520,46 @@ func (e *Endpoint) Update(owner Owner, opts models.ConfigurationMap) error {
 	}
 
 	if needToRegenerate {
-		stateTransitionSucceeded := e.SetStateLocked(StateWaitingToRegenerate, reason)
-
-		e.Mutex.Unlock()
-		ctCleaned.Wait()
-
-		if stateTransitionSucceeded {
-			e.Regenerate(owner, reason)
-			return nil
+		scopedLog.Debug("need to regenerate endpoint; checking state before" +
+			" attempting to regenerate")
+
+		// TODO / FIXME: GH-3281: need ways to queue up regenerations per-endpoint.
+
+		// Default timeout for PATCH /endpoint/{id}/config is 30 seconds, so put
+		// timeout in this function a bit below that timeout. If the timeout
+		// for clients in API is below this value, they will get a message containing
+		// "context deadline exceeded".
+		stateChangeTimeout := time.Duration(25 * time.Second)
+
+		// Check up until stateChangeTimeout seconds for endpoint state before
+		// trying to update configuration.
+		timeout := time.After(stateChangeTimeout)
+
+		// Check for endpoint state every second.
+		ticker := time.NewTicker(1 * time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				// Check endpoint state before attempting configuration update because
+				// configuration updates can only be applied when the endpoint is in
+				// specific states. See GH-3058.
+				stateTransitionSucceeded := e.SetStateLocked(StateWaitingToRegenerate, reason)
```

Comment:
I don't think this can work as you keep the endpoint locked so it's impossible for it to get regenerated or the identity can be set as that requires taking the endpoint lock. I don't see how the state can ever change if a build is queued.

We need to unlock and relock after the sleep interval.

Reply:
Good catch.